### PR TITLE
Introduce a locally runnable matching simulator

### PR DESCRIPTION
**Pull request by taylanisikdemir**
**What changed?**
Partitioned tasklist matching logic could use some improvements in order to match the tasks with pollers as soon as possible. We observed some sub-optimal behavior which becomes bottleneck for heavy traffic partitioned tasklists. Especially with isolation groups feature enabled. Therefore we are revisiting matching partitioning/routing/forwarding/isolation features. 

This PR introduces a locally runnable matching simulation with various knobs and collects/reports useful information via a new structured event logging. 

Simulator configuration looks like this:
```
matchingconfig:
  nummatchinghosts: 4
  simulationconfig:
    tasklistwritepartitions: 2
    tasklistreadpartitions: 2
    numpollers: 10
    numtaskgenerators: 2
    taskgeneratortickinterval: 50ms
    maxtasktogenerate: 1500
    polltimeout: 5s
    forwardermaxoutstandingpolls: 20
    forwardermaxoutstandingtasks: 1
    forwardermaxratepersecond: 10
    forwardermaxchildrenpernode: 20
```

The structured event logs are added all over the matching code for better visibility. These event logs are only printed when `MATCHING_LOG_EVENTS=true` so they will only be available for simulation runs for now. Example event look like this:
```
{
  "DomainID": "2d84398e-1a5b-4436-86b3-4d6c86a0aa72",
  "WorkflowID": "test-workflow-id",
  "RunID": "ccb6ecce-e90c-455f-93d3-8564e6ff298d",
  "TaskID": 8,
  "ScheduleID": 21,
  "CreatedTime": "2024-07-30T22:48:55.539Z",
  "PartitionConfig": null,
  "TaskListName": "my-tasklist",
  "TaskListKind": "NORMAL",
  "TaskListType": 0,
  "EventTime": "2024-07-30T22:48:55.544875679Z",
  "EventName": "Matched Task (pollOrForward)",
  "Host": "",
  "Payload": {
    "FromIsolatedTaskC": false,
    "IsolationGroup": "",
    "SyncMatched": false,
    "TaskIsForwarded": false
  }
}
```

The main simulator logic is in `host/matching_simulation_test.go`. 
Rest of the changes are updates to integration test framework to satisfy the simulator's needs such as
- Ability to run multiple matching services
- Ability to mock history client

**How did you test it?**
```
./scripts/run_matching_simulator.sh
```

This script runs the test for 1 minute and prints a simulation summary by querying the event logs via jq. 

Example simulation summary output: https://gist.github.com/taylanisikdemir/7cebe12fdf6aaf72e0005bd13bbec7e0

One weird thing (which I hope is a bug in my setup) is that sync matches only happen for tasks forwarded to root partition. This needs to be debugged a bit to make sure the setup is not acting weird.

**Next steps**
- Instrument other/missed critical places with event logs
- Run simulation with different configurations such as
    - high vs low partition count
    - mismatching read/write partition counts
    - forwarding tasks enabled vs disabled
    - forwarding polls enabled vs disabled
    - generate more tasks and run more pollers
- Some ideas to try
    - Sleep a bit to give sync match a chance before forwarding
    - Forward polls/tasks to a random partition instead of root one

### CODE REVIEW

File: service/matching/handler/engine.go [hunk 1]
```diff
@@ -441,6 +466,16 @@ func (e *matchingEngineImpl) PollForDecisionTask(
...
 	)
+	event.Log(event.E{
+		TaskListName: taskListName,
+		TaskListKind: taskListKind,
+		TaskListType: persistence.TaskListTypeDecision,
+		TaskInfo: persistence.TaskInfo{
+			DomainID: domainID,
+		},
+		EventName: "Received PollForDecisionTask",
+		Host:      e.config.HostName,
```
Comment (Shaddoll) [inline service/matching/handler/engine.go:477]: do we need to include `request.GetForwardedFrom()` in the payload?

File: service/matching/handler/engine.go [hunk 2]
```diff
@@ -540,7 +596,22 @@ pollLoop:
...
 		task.Finish(nil)
+		event.Log(event.E{
+			TaskListName: taskListName,
+			TaskListKind: taskListKind,
+			TaskListType: persistence.TaskListTypeDecision,
+			TaskInfo:     task.Info(),
+			EventName:    "PollForDecisionTask returning task",
+			Host:         e.config.HostName,
+			Payload: map[string]any{
+				"TaskIsForwarded":      task.IsForwarded(),
```
Comment (Shaddoll) [inline service/matching/handler/engine.go:609]: we can include the source of task (whether from db or history)

**Review by Shaddoll — APPROVED** [review]

File: service/matching/handler/engine.go (hunk 1 above)
Comment (taylanisikdemir) [reply service/matching/handler/engine.go:477]: added.

File: service/matching/handler/engine.go (hunk 2 above)
Comment (taylanisikdemir) [reply service/matching/handler/engine.go:612]: good idea. I'll leave that to a follow up PR

**Pull request merged**